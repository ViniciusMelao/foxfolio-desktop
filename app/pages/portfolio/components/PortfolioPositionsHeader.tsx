import { Button, Card, CardContent, Grid, Typography } from '@material-ui/core';
import { StyleRulesCallback, Theme, withStyles } from '@material-ui/core/styles';
import React from 'react';

const styles: StyleRulesCallback = (theme: Theme) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
    paddingTop: '12px !important',
    paddingBottom: '12px !important',
  },
  avatar: {
    flex: '0 0 auto',
    width: 40,
    marginRight: theme.spacing.unit * 2,
  },
  content: {
    flex: '1 1 auto',
    marginRight: theme.spacing.unit * 2,
  },
  collapseButton: {
    flex: '0 0 auto',
    width: 48,
  },
  right: {
    textAlign: 'right',
  },
  interval: {
    fontSize: 12,
    color: 'grey',
  },
});

export const PortfolioPositionHeader = withStyles(styles)(({ classes }) => (
  <Card square>
    <CardContent className={classes.root}>
      <div className={classes.avatar} />
      <div className={classes.content}>
        <Grid container>
          <Grid item xs={4}>
            <Typography variant="subheading">Coin</Typography>
          </Grid>
          <Grid item xs={3} className={classes.right}>
            <Typography variant="subheading">Quantity</Typography>
          </Grid>
          <Grid item xs={3} className={classes.right}>
            <Typography variant="subheading">Price</Typography>
          </Grid>
          <Grid item xs={2} className={classes.right}>
            <Typography variant="subheading">Change</Typography>
          </Grid>
        </Grid>
      </div>
      <div className={classes.collapseButton} />
    </CardContent>
  </Card>
));
