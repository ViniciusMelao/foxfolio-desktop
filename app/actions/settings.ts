import { CurrencyFocus, SettingsType } from '../modules/settings.types';
import { getSettings } from '../selectors/selectGlobalState';
import { Action, Dispatch, GetState, ThunkAction } from './actions.types';
import { requestTickerUpdate } from './ticker';

function saveSettings(settings: SettingsType): Action {
  return {
    type: 'SAVE_SETTINGS',
    settings,
  };
}

export function setPortfolioFocus(currencyFocus: CurrencyFocus): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(saveSettings({ ...getSettings(getState()), currencyFocus }));
  };
}

export function saveSettingsAndUpdateTicker(settings: SettingsType): ThunkAction {
  return (dispatch: Dispatch) => {
    dispatch(saveSettings(settings));
    dispatch(requestTickerUpdate([], settings.fiatCurrency, settings.cryptoCurrency));
  };
}
