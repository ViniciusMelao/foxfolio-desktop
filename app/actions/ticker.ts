import 'whatwg-fetch'; // Has to be imported before ccxt

import * as ccxt from 'ccxt';
import dayjs from 'dayjs';
import _ from 'lodash';
import { getHistoryEntry, getTickerEntry } from '../helpers/ticker';
import { Coinlist } from '../modules/coinlist.types';
import { Exchange, Exchanges } from '../modules/exchanges.types';
import { HistoryData, Ticker, TickerEntry } from '../modules/ticker.types';
import { Wallet } from '../modules/wallets.types';
import { getExchanges, getHistory } from '../selectors/selectGlobalState';
import { Action, Dispatch, GetState, ThunkAction } from './actions.types';
import startTimer from './timer';

const REFRESH_TIME_IN_MS = 30000;

function fetchingTickerUpdate(): Action {
  return {
    type: 'LAST_UPDATED',
    key: 'ticker',
    time: new Date(),
  };
}

function receiveTickerUpdate(rawTicker: RawTicker): Action {
  return {
    type: 'TICKER_UPDATE',
    ticker: formatRawTicker(rawTicker),
  };
}

function receiveHistory(fsym: string, tsym: string, history: HistoryData): Action {
  return {
    type: 'HISTORY_UPDATE',
    fsym,
    tsym,
    history,
  };
}

function receiveCoinList(coinlist: Coinlist): Action {
  return {
    type: 'RECEIVE_COIN_LIST',
    coinlist,
  };
}

export function requestTickerUpdate(
  extraSymbols: string[] = [],
  fiatCurrency?: string,
  cryptoCurrency?: string
): ThunkAction {
  return async (dispatch: Dispatch, getState: GetState) => {
    dispatch(fetchingTickerUpdate());

    const state = getState();
    fiatCurrency = fiatCurrency || state.settings.fiatCurrency;
    cryptoCurrency = cryptoCurrency || state.settings.cryptoCurrency;

    const symbols = getSymbolsFromTransactions(
      state.exchanges,
      state.wallets,
      fiatCurrency,
      cryptoCurrency,
      extraSymbols
    );
    try {
      let ticker = await getTickerForSymbols(symbols);
      ticker = calculateFiatValues(
        ticker,
        fiatCurrency,
        cryptoCurrency,
        getTickerEntry(ticker, cryptoCurrency, fiatCurrency)
      );
      ticker = await requestMissingTickerUpdateFromExchange(ticker, getExchanges(state), symbols);
      dispatch(receiveTickerUpdate(ticker));
    } catch (error) {
      console.error(error); // TODO Show this error to the user? Logging?
    }
  };
}

const getTickerForSymbols = (symbols: Symbols) => {
  const fsyms = symbols.from.join(',');
  const tsyms = symbols.to.join(',');
  if (fsyms && tsyms) {
    return fetch(
      `https://min-api.cryptocompare.com/data/pricemultifull?fsyms=${fsyms}&tsyms=${tsyms}`
    )
      .then(result => result.json())
      .then(result => result.RAW);
  }
  return Promise.reject('No symbols');
};

const calculateChange = (tickerEntry: any, tickerEntryFiat: any): number => {
  const open24hour = tickerEntry.OPEN24HOUR * tickerEntryFiat.OPEN24HOUR;
  const currentPrice = tickerEntry.PRICE * tickerEntryFiat.PRICE;

  return (currentPrice - open24hour) / open24hour * 100;
};

const calculateFiatValues = (
  ticker: Ticker,
  fiatCurrency: string,
  cryptoCurrency: string,
  tickerEntryFiatCrypto: TickerEntry
): Ticker => {
  return _.mapValues(ticker, (entry, asset) => {
    if (asset === fiatCurrency || asset === cryptoCurrency) {
      return entry;
    }
    return {
      ...entry,
      [fiatCurrency]: {
        ...(entry ? entry[fiatCurrency] : {}),
        CHANGEPCT24HOUR: calculateChange(
          getTickerEntry(ticker, asset, cryptoCurrency),
          tickerEntryFiatCrypto
        ),
      },
    };
  });
};

const requestMissingTickerUpdateFromExchange = async (
  ticker: Ticker,
  exchanges: Exchanges,
  symbols: Symbols
) => {
  const updatedTicker: Ticker = { ...ticker };

  try {
    for (const fsym of symbols.from) {
      if (!ticker[fsym]) {
        for (const exchange of _.values(exchanges)) {
          if (exchangeHasSymbol(fsym, exchange)) {
            const connector: ccxt.Exchange = new ccxt[exchange.type](exchange.credentials);
            const markets: {
              [symbol: string]: ccxt.Market;
            } = (await connector.loadMarkets()) as any;
            const marketsForSymbol = _.values(markets).filter(market =>
              market.symbol.startsWith(fsym.toUpperCase())
            );
            for (const market of marketsForSymbol) {
              try {
                const tick = await connector.fetchTicker(market.symbol);
                updatedTicker[market.base] = {
                  ...updatedTicker[market.base],
                  [market.quote]: {
                    PRICE: tick.last || tick.ask,
                    CHANGEPCT24HOUR: 0,
                  },
                };
              } catch (error) {
                console.error(error);
              }
            }
            break;
          }
        }
      }
    }
  } catch (error) {
    console.error(error);
    return ticker;
  }
  return updatedTicker;
};

const exchangeHasSymbol = (symbol: string, exchange: Exchange) => {
  return _.includes(_.keys(exchange.balances), symbol);
};

export function requestHistory(
  fsym: string,
  tsym: string,
  forceRequest: boolean = false
): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    const twoMinutesAgo = dayjs().subtract(2, 'minute');
    if (
      forceRequest ||
      !(getHistoryEntry(getHistory(getState()), fsym, tsym).lastUpdate >= twoMinutesAgo.toDate())
    ) {
      fetch(`https://min-api.cryptocompare.com/data/histominute?fsym=${fsym}&tsym=${tsym}`)
        .then(result => result.json())
        .then(result => dispatch(receiveHistory(fsym, tsym, result.Data)))
        .catch(error => console.error(error));
    }
  };
}

export const requestHistoryForAll = (
  fsyms: string[],
  tsym: string,
  forceRequest: boolean = false
) => (dispatch: Dispatch) =>
  fsyms.forEach(fsym => dispatch(requestHistory(fsym, tsym, forceRequest)));

export function requestCoinList(): ThunkAction {
  return (dispatch: Dispatch) => {
    fetch('https://min-api.cryptocompare.com/data/all/coinlist')
      .then(result => result.json())
      .then(result => dispatch(receiveCoinList(result.Data)))
      .catch(error => console.error(error));
  };
}

export function continuouslyUpdateTicker(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    if (!getState().timers.timers.ticker) {
      const timer = window.setInterval(() => dispatch(requestTickerUpdate()), REFRESH_TIME_IN_MS);
      dispatch(startTimer('ticker', timer));
    }
  };
}

interface Symbols {
  from: string[];
  to: string[];
}

function getSymbolsFromTransactions(
  exchanges: Exchanges,
  wallets: Wallet[],
  fiatCurrency: string,
  cryptoCurrency: string,
  extraSymbols: string[]
): Symbols {
  const walletSymbols = wallets.map(wallet => wallet.currency) || [];
  const exchangeSymbols = _.chain(exchanges)
    .values()
    .map(exchange => exchange.balances)
    .map(_.keys)
    .reduce((acc, keys) => acc.concat(keys), [] as string[])
    .value();

  return {
    from: _.uniq([cryptoCurrency, ...walletSymbols, ...exchangeSymbols, ...extraSymbols]),
    to: [cryptoCurrency, fiatCurrency],
  };
}

interface RawTicker {
  [fsym: string]: {
    [tsym: string]: RawTickerEntry;
  };
}

interface RawTickerEntry {
  CHANGEPCT24HOUR: string;
  PRICE: string;
}

function formatRawTicker(rawTicker: RawTicker): Ticker {
  return _.mapValues(rawTicker, fsymEntry =>
    _.mapValues(fsymEntry, (entry: RawTickerEntry) => ({
      ...entry,
      CHANGEPCT24HOUR: parseFloat(entry.CHANGEPCT24HOUR),
      PRICE: parseFloat(entry.PRICE),
    }))
  );
}
