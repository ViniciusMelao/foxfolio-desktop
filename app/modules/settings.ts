import { Action } from '../actions/actions.types';
import { SettingsType } from './settings.types';

export const MINIMUM_BALANCE = 0.05;

export const initialSettings: SettingsType = {
  fiatCurrency: 'USD',
  cryptoCurrency: 'BTC',
  hideZeroBalances: false,
  includeFiat: true,
  currencyFocus: 'fiat',
  theme: 'light',
};

export default function settings(
  state: SettingsType = initialSettings,
  action: Action
): SettingsType {
  switch (action.type) {
    case 'SAVE_SETTINGS':
      return action.settings;
    default:
      return state;
  }
}
