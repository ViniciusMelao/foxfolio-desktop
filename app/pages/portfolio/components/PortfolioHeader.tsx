import { Grid, Typography } from '@material-ui/core';
import React from 'react';
import { SettingsType } from '../../../modules/settings.types';
import { PortfolioChange, PortfolioSum } from '../types/portfolio.types';
import { PriceChangeText } from './PriceChangeText';

interface Props {
  settings: SettingsType;
  sum: PortfolioSum;
  change: PortfolioChange;
}

export const PortfolioHeader = ({ settings, sum, change }: Props) => {
  return (
    <Grid container>
      <Grid item xs={2} />
      <Grid item xs={4}>
        <Typography variant="title">
          {sum.fiat.toPrecision(5)} {settings.fiatCurrency} |&nbsp;
          <PriceChangeText change={change.fiat} />
        </Typography>
      </Grid>
      <Grid item xs={4}>
        <Typography variant="title" style={{ paddingLeft: 20 }}>
          {sum.crypto.toPrecision(5)} {settings.cryptoCurrency} |{' '}
          <PriceChangeText change={change.crypto} />
        </Typography>
      </Grid>
    </Grid>
  );
};
