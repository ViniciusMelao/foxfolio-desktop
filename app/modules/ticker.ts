import { Action } from '../actions/actions.types';
import { TickerAndHistory } from './ticker.types';

export default function reducer(
  state: TickerAndHistory = { ticker: {}, history: {}, pricesForTime: {} },
  action: Action
): TickerAndHistory {
  switch (action.type) {
    case 'TICKER_UPDATE':
      return { ...state, ticker: action.ticker };
    case 'HISTORY_UPDATE':
      return {
        ...state,
        history: {
          ...state.history,
          [action.fsym]: {
            ...state.history[action.fsym],
            [action.tsym]: {
              lastUpdate: new Date(),
              data: action.history.filter((val, i) => i % 5 === 0),
            },
          },
        },
      };
    case 'RECEIVE_PRICE_FOR_TIME':
      const pricesForSymbol = state.pricesForTime[action.fsym];
      return {
        ...state,
        pricesForTime: {
          ...state.pricesForTime,
          [action.fsym]: {
            ...(state.pricesForTime ? state.pricesForTime[action.fsym] : {}),
            [action.tsym]: {
              ...(pricesForSymbol ? pricesForSymbol[action.tsym] : {}),
              [action.timestamp]: action.price,
            },
          },
        },
      };
    default:
      return state;
  }
}
