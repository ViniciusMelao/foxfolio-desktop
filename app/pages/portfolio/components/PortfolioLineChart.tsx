import { LinearProgress, withTheme } from '@material-ui/core';
import { WithTheme } from '@material-ui/core/styles/withTheme';
import dayjs from 'dayjs';
import React, { Component } from 'react';
import { connect, MapStateToProps } from 'react-redux';
import {
  Area,
  AreaChart,
  CartesianGrid,
  ResponsiveContainer,
  Tooltip,
  XAxis,
  YAxis,
} from 'recharts';
import { bindActionCreators } from 'redux';

import * as _ from 'lodash';
import { Dispatch } from '../../../actions/actions.types';
import { getHistoryEntry } from '../../../helpers/ticker';
import { GlobalState } from '../../../modules';
import * as HistoryActions from '../../../modules/history';
import { History } from '../../../modules/history.types';
import { SettingsType } from '../../../modules/settings.types';
import { getHistory, getSettings } from '../../../selectors/selectGlobalState';
import { Balances } from '../types/portfolio.types';

interface StateProps {
  history: History;
  settings: SettingsType;
}

interface DispatchProps {
  requestHistoryForAll: (fsyms: string[], tsym: string) => void;
}

interface OwnProps {
  portfolio: Balances;
  sum: number;
}

export const ThemedPortfolioLineChart = withTheme()(
  class extends Component<StateProps & DispatchProps & OwnProps & WithTheme> {
    public componentWillMount() {
      const { requestHistoryForAll, portfolio, settings } = this.props;
      requestHistoryForAll(Object.keys(portfolio), settings.cryptoCurrency);
    }

    public render() {
      const { history, portfolio, sum, settings, theme } = this.props;

      const data = calculateHistoryForPortfolio(history, portfolio, settings.cryptoCurrency, sum);
      if (data.length > 0) {
        return (
          <ResponsiveContainer height={300}>
            <AreaChart data={data} margin={{ left: 5, top: 5, bottom: 0, right: 15 }}>
              <Area
                stroke={theme.palette.text.secondary}
                fill={theme.palette.divider}
                dataKey="value"
                isAnimationActive={false}
              />
              <XAxis
                dataKey="time"
                tick={{ fill: theme.palette.text.secondary, strokeWidth: 0 }}
                tickFormatter={tick => dayjs.unix(tick).format('HH:mm')}
              />
              <YAxis
                interval="preserveStartEnd"
                tick={{ fill: theme.palette.text.secondary, strokeWidth: 0 }}
                tickFormatter={tick => tick.toPrecision(3)}
                domain={[dataMin => dataMin - 0.02 * dataMin, dataMax => dataMax + 0.02 * dataMax]}
              />
              <CartesianGrid stroke={theme.palette.divider} strokeDasharray="3 3" />
              <Tooltip
                itemStyle={{ color: theme.palette.text.secondary }}
                wrapperStyle={{ backgroundColor: theme.palette.background.paper }}
                labelFormatter={() => ''}
              />
            </AreaChart>
          </ResponsiveContainer>
        );
      } else {
        return <LinearProgress />;
      }
    }
  }
);

const calculateHistoryForPortfolio = (
  history: History,
  portfolio: Balances,
  tsym: string,
  sum: number
): Array<{ time: number; value: number }> => {
  _.mapValues(portfolio, balance => balance / sum);
  const mergedHistory: { [time: string]: number } = {};
  Object.keys(portfolio).forEach(fsym => {
    const data = getHistoryEntry(history, fsym, tsym).data;
    data.forEach(
      value =>
        (mergedHistory[value.time] =
          (mergedHistory[value.time] || 0) + value.close * portfolio[fsym])
    );
  });
  return _.entries(mergedHistory).map(pair => ({
    time: parseInt(pair[0], 10),
    value: Math.floor((pair[1] + (portfolio[tsym] || 0)) * 1e6) / 1e6,
  }));
};

const mapStateToProps: MapStateToProps<StateProps, OwnProps, GlobalState> = state => ({
  history: getHistory(state),
  settings: getSettings(state),
});
const mapActionsToProps = (dispatch: Dispatch) => bindActionCreators(HistoryActions, dispatch);

export const PortfolioLineChart = connect<StateProps, DispatchProps, OwnProps, GlobalState>(
  mapStateToProps,
  mapActionsToProps
)(ThemedPortfolioLineChart);
