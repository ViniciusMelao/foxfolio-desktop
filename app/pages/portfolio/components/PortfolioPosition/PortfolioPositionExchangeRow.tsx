import { Avatar, Card, CardContent, Grid, Typography, withStyles } from '@material-ui/core';
import { StyleRulesCallback } from '@material-ui/core/styles';
import { TrendingUp } from '@material-ui/icons';
import React from 'react';
import { connect } from 'react-redux';
import { GlobalState } from '../../../../modules';
import { Exchanges } from '../../../../modules/exchanges.types';
import { getExchanges } from '../../../../selectors/selectGlobalState';

const styles: StyleRulesCallback = theme => ({
  root: {
    display: 'flex',
    alignItems: 'center',
  },
  avatar: {
    backgroundColor: theme.palette.text.secondary,
    flex: '0 0 auto',
    marginRight: theme.spacing.unit * 2,
    margin: 5,
    width: 30,
    height: 30,
  },
  content: {
    flex: '1 1 auto',
  },
  right: {
    textAlign: 'right',
  },
});

interface Props {
  asset: string;
  balance: number;
  exchangeKey: string;
  exchanges: Exchanges;
}

const styledComponent = withStyles(styles)<Props>(
  ({ asset, balance, classes, exchangeKey, exchanges }) => (
    <Card>
      <CardContent className={classes.root}>
        <div>
          <Avatar className={classes.avatar}>
            <TrendingUp />
          </Avatar>
        </div>
        <div className={classes.content}>
          <Grid container>
            <Grid item xs={3}>
              <Typography variant="body2" component="span">
                {exchanges[exchangeKey].type}
              </Typography>
            </Grid>
            <Grid item xs={2} className={classes.right}>
              <Typography>{`${balance.toPrecision(5)} ${asset}`}</Typography>
            </Grid>
          </Grid>
        </div>
        <div style={{ width: 48 }} />
      </CardContent>
    </Card>
  )
);

const mapStateToProps = (state: GlobalState) => ({
  exchanges: getExchanges(state),
});

export const PortfolioPositionExchangeRow = connect(mapStateToProps)(styledComponent);
