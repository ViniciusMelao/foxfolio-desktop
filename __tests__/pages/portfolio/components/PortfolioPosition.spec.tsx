import { createShallow } from '@material-ui/core/test-utils';
import * as React from 'react';
import { initialSettings } from '../../../../app/modules/settings.types';
import { Ticker } from '../../../../app/modules/ticker.types';
import { PortfolioPosition } from '../../../../app/pages/portfolio/components/PortfolioPosition';
import { FiatCurrency } from '../../../../app/utils/currencies';

let shallow;

beforeAll(() => {
  shallow = createShallow();
});

let coinlist = {};
let ticker: Ticker;
const settings = { ...initialSettings, fiatCurrency: 'USD' as FiatCurrency };

beforeEach(() => {
  coinlist = {};
  ticker = {
    BTC: { USD: { PRICE: 10000, CHANGEPCT24HOUR: 1.1 } },
    ETH: {
      BTC: { PRICE: 0.1, CHANGEPCT24HOUR: 2 },
      USD: { PRICE: 1000, CHANGEPCT24HOUR: 5.2 },
    },
  };
});

test('Render portfolio position', () => {
  const asset = 'ETH';
  const portfolio = {
    total: 10,
    wallets: 2,
    exchanges: { exchangeKey: 8 },
  };

  const wrapper = shallow(
    <PortfolioPosition
      asset={asset}
      coinlist={coinlist}
      portfolio={portfolio}
      settings={settings}
      ticker={ticker}
    />
  ).dive();
  expect(wrapper).toMatchSnapshot();
});

test('Portfolio position with missing ticker', () => {
  const asset = 'ETH';
  const portfolio = {
    total: 10,
    wallets: 2,
    exchanges: { exchangeKey: 8 },
  };

  ticker = {
    BTC: { BTC: { PRICE: 1, CHANGEPCT24HOUR: 0 }, USD: { PRICE: 10000, CHANGEPCT24HOUR: 1.1 } },
    ETH: {
      USD: { PRICE: 1000, CHANGEPCT24HOUR: 5.2 },
    },
  };

  const wrapper = shallow(
    <PortfolioPosition
      asset={asset}
      coinlist={coinlist}
      portfolio={portfolio}
      settings={settings}
      ticker={ticker}
    />
  ).dive();
  expect(wrapper).toMatchSnapshot();
});
