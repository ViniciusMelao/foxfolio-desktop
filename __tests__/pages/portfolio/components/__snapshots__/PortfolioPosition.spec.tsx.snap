// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portfolio position with missing ticker 1`] = `
<WithStyles(class_1)
  cardContent={
    <React.Fragment>
      <div
        className="Component-avatar-1"
      >
        <WithStyles(Component)
          asset="ETH"
          coinlist={Object {}}
        />
      </div>
      <div
        className="Component-content-2"
      >
        <WithStyles(Grid)
          container={true}
        >
          <WithStyles(Grid)
            item={true}
            xs={4}
          >
            <div>
              <WithStyles(Typography)
                color="default"
                component="span"
                variant="body2"
              >
                ETH
              </WithStyles(Typography)>
              <WithStyles(Typography)
                color="textSecondary"
                component="span"
                variant="body2"
              >
                10.000
              </WithStyles(Typography)>
            </div>
          </WithStyles(Grid)>
          <WithStyles(Grid)
            className="Component-right-3"
            item={true}
            xs={3}
          >
            
          </WithStyles(Grid)>
          <WithStyles(Grid)
            className="Component-right-3"
            item={true}
            xs={3}
          >
            
          </WithStyles(Grid)>
          <WithStyles(Grid)
            className="Component-right-3"
            item={true}
            xs={2}
          >
            
          </WithStyles(Grid)>
        </WithStyles(Grid)>
      </div>
    </React.Fragment>
  }
  collapseContent={
    <Connect(WithTheme(class_1))
      fsym="ETH"
      tsym="USD"
    />
  }
/>
`;

exports[`Render portfolio position 1`] = `
<WithStyles(class_1)
  cardContent={
    <React.Fragment>
      <div
        className="Component-avatar-1"
      >
        <WithStyles(Component)
          asset="ETH"
          coinlist={Object {}}
        />
      </div>
      <div
        className="Component-content-2"
      >
        <WithStyles(Grid)
          container={true}
        >
          <WithStyles(Grid)
            item={true}
            xs={4}
          >
            <div>
              <WithStyles(Typography)
                color="default"
                component="span"
                variant="body2"
              >
                ETH
              </WithStyles(Typography)>
              <WithStyles(Typography)
                color="textSecondary"
                component="span"
                variant="body2"
              >
                10.000
              </WithStyles(Typography)>
            </div>
          </WithStyles(Grid)>
          <WithStyles(Grid)
            className="Component-right-3"
            item={true}
            xs={3}
          >
            <div
              style={
                Object {
                  "alignItems": "center",
                  "display": "flex",
                  "height": "100%",
                  "justifyContent": "flex-end",
                }
              }
            >
              <WithStyles(Typography)
                color="default"
                component="span"
                style={
                  Object {
                    "padding": "auto",
                  }
                }
                variant="subheading"
              >
                10,000.00 USD
              </WithStyles(Typography)>
              
            </div>
          </WithStyles(Grid)>
          <WithStyles(Grid)
            className="Component-right-3"
            item={true}
            xs={3}
          >
            <div
              style={
                Object {
                  "alignItems": "center",
                  "display": "flex",
                  "height": "100%",
                  "justifyContent": "flex-end",
                }
              }
            >
              <WithStyles(Typography)
                color="default"
                component="span"
                style={
                  Object {
                    "padding": "auto",
                  }
                }
                variant="subheading"
              >
                1,000.00 USD
              </WithStyles(Typography)>
              
            </div>
          </WithStyles(Grid)>
          <WithStyles(Grid)
            className="Component-right-3"
            item={true}
            xs={2}
          >
            <div
              style={
                Object {
                  "alignItems": "center",
                  "display": "flex",
                  "height": "100%",
                  "justifyContent": "flex-end",
                }
              }
            >
              <WithStyles(Typography)
                color="default"
                component="span"
                style={
                  Object {
                    "padding": "auto",
                  }
                }
                variant="subheading"
              >
                <WithStyles(Component)
                  change={5.2}
                  muted={false}
                />
              </WithStyles(Typography)>
              
            </div>
          </WithStyles(Grid)>
        </WithStyles(Grid)>
      </div>
    </React.Fragment>
  }
  collapseContent={
    <Connect(WithTheme(class_1))
      fsym="ETH"
      tsym="USD"
    />
  }
/>
`;
